modules/chemical_reactions/src/auxkernels/AqueousEquilibriumRxnAux.C

```diff
@@ -16,19 +16,40 @@ validParams<AqueousEquilibriumRxnAux>()
                                              "The stoichiometric coefficient of reactants");
   params.addCoupledVar("v",
                        "The list of primary species participating in this equilibrium species");
+  params.addCoupledVar("gamma", 1.0, "Activity coefficient of this secondary equilibrium species");
+  params.addCoupledVar("gamma_v", 1.0, "Activity coefficients of coupled primary species");
   params.addClassDescription("Concentration of secondary equilibrium species");
   return params;
 }
 
 AqueousEquilibriumRxnAux::AqueousEquilibriumRxnAux(const InputParameters & parameters)
   : AuxKernel(parameters),
     _log_k(getParam<Real>("log_k")),
-    _sto_v(getParam<std::vector<Real>>("sto_v"))
+    _sto_v(getParam<std::vector<Real>>("sto_v")),
+    _gamma_eq(coupledValue("gamma"))
 {
-  int n = coupledComponents("v");
+  const unsigned int n = coupledComponents("v");
+
+  // Check that the correct number of stoichiometric coefficients have been provided
+  if (_sto_v.size() != n)
+    mooseError("The number of stoichiometric coefficients and coupled species must be equal in ",
+               _name);
+
+  // Check that the correct number of activity coefficients have been provided (if applicable)
+  if (isCoupled("gamma_v"))
+    if (coupled("gamma_v") != n)
+      mooseError("The number of activity coefficients and coupled species must be equal in ",
+                 _name);
+
   _vals.resize(n);
-  for (unsigned int i = 0; i < _vals.size(); ++i)
+  _gamma_v.resize(n);
+
+  for (unsigned int i = 0; i < n; ++i)
+  {
     _vals[i] = &coupledValue("v", i);
+    // If gamma_v has been supplied, use those values, but if not, use the default value
+    _gamma_v[i] = (isCoupled("gamma_v") ? &coupledValue("gamma_v", i) : &coupledValue("gamma_v"));
```

Comment:
Hopefully this isn't too cute. To avoid lots of `if (isCoupled("gamma_v")` statements in the residual and jacobians, I wanted `_gamma_v` to be a vector of VariableValues when it is coupled, and a vector of VariableValues equal to one (the default value) when it isn't.

Reply:
👍